### PR TITLE
better token count management

### PR DESCRIPTION
**Pull request by jerpint**
In this PR, we finally check if documents are long enough or not using token counts instead of word counts. 

If a collecrtion of documents is too long, it will truncate as much as it can fit, and return all necessary documents.

### CODE REVIEW

File: buster/tokenizers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from .base import Tokenizer
+from .gpt import GPTTokenizer
+
+
+def tokenizer_factory(tokenizer_cfg):
+    print(tokenizer_cfg)
```
Comment (hbertrand) [inline buster/tokenizers/__init__.py:6]: leftover print

File: buster/tokenizers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from .base import Tokenizer
+from .gpt import GPTTokenizer
+
+
+def tokenizer_factory(tokenizer_cfg):
```
Comment (hbertrand) [inline buster/tokenizers/__init__.py:5]: typehints

File: buster/tokenizers/base.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    @abstractmethod
+    def encode(self, string: str) -> list[int]:
+        ...
+
+    @abstractmethod
+    def decode(self, encoded: list[int]):
+        ...
+
+    def num_tokens(self, string: str, return_encoded: bool = False):
```
Comment (hbertrand) [inline buster/tokenizers/base.py:18]: return typehint

File: buster/busterbot.py [hunk 4]
```diff
@@ -88,13 +95,20 @@ def update_cfg(self, cfg: BusterConfig):
...
         self.unk_embedding = self.get_embedding(self.unknown_prompt, engine=self.embedding_model)
 
         # update completer and formatter cfg
+        self.tokenizer = tokenizer_factory(self.tokenizer_cfg)
         self.completer = completer_factory(self.completion_cfg)
-        self.prompt_formatter = prompt_formatter_factory(self.prompt_cfg)
+        self.documents_formatter = document_formatter_factory(
+            tokenizer=self.tokenizer,
+            max_tokens=self.retriever_cfg["max_tokens"]
+            # TODO: move max_tokens from retriever_cfg to somewhere more logical
```
Comment (hbertrand) [inline buster/busterbot.py:109]: wait it seems to be in `prompt_cfg` but not `retriever_cfg`, looking at lines 41-49. How is this not crashing?

File: buster/completers/base.py [hunk 5]
```diff
@@ -43,6 +43,9 @@ def generate_response(self, system_prompt, user_input) -> Completion:
         logger.info(f"{user_input=}")
         try:
             completion = self.complete(prompt=prompt, **self.completion_kwargs)
+        except openai.error.InvalidRequestError:
+            logger.exception("Invalid request to OpenAI API. See traceback:")
+            return Completion("Something went wrong, try again soon!", True, "Invalid request made to openai.")
```
Comment (hbertrand) [inline buster/completers/base.py:48]: overlap with the previous PR?

File: buster/formatters/documents.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        if num_preserved_docs < (num_total_docs):
+            logger.warning(
+                f"{num_preserved_docs}/{num_total_docs} documents were preserved from the matched documents due to truncation."
+            )
+            matched_documents = matched_documents.iloc[:num_preserved_docs]
+
+        return documents_str, matched_documents
+
+
+def document_formatter_factory(tokenizer: Tokenizer, max_tokens):
```
Comment (hbertrand) [inline buster/formatters/documents.py:54]: kinda overkill at this point

File: tests/test_formatters.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    document_1 = "This is a very short document."
+    document_2 = "This is another very short document."
+    document_3 = "This is also a short document."
+
+    matched_documents = pd.DataFrame({"content": [document_1, document_2, document_3]})
+
+    docs_str, matched_documents_new = documents_formatter.format(matched_documents)
+
+    # less documents and the new document is shorter than the original
+    matched_documents.content == matched_documents_new.content
```
Comment (hbertrand) [inline tests/test_formatters.py:28]: there is no assert, is this actually doing something ?

File: tests/test_formatters.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+def test_documents_formatter__normal():
+    tokenizer = GPTTokenizer(model_name="gpt-3.5-turbo")
+    documents_formatter = DocumentsFormatter(
+        tokenizer=tokenizer,
+        max_tokens=100,
+    )
+
+    # In this test, document_1 doesn't entirely fit.
+    # we only expect a part of it to be contained.
```
Comment (hbertrand) [inline tests/test_formatters.py:17]: incorrect comment

**Review by hbertrand — CHANGES_REQUESTED** [review]

File: buster/completers/base.py (hunk 5 above)
Comment (jerpint) [reply buster/completers/base.py:48]: yes i forgot this bit from the last PR

File: tests/test_formatters.py (hunk 7 above)
Comment (jerpint) [reply tests/test_formatters.py:28]: youre right, it wasnt, added an `assert all()`

File: buster/formatters/documents.py (hunk 6 above)
Comment (jerpint) [reply buster/formatters/documents.py:54]: yes but its consistent with the rest of the code, though i think at some point we may want to drop these factories all together

File: buster/busterbot.py (hunk 4 above)
Comment (jerpint) [reply buster/busterbot.py:109]: it should be in both, where are you looking?  We have a max_tokens for the total prompt, but also a max_tokens for the sum of all documents

File: buster/completers/base.py (hunk 5 above)
Comment (jerpint) [reply buster/completers/base.py]: these configs are also starting to get disorganized... might be time to switch over to omegaconf or to do a rehaul

File: buster/tokenizers/base.py [hunk 9]
```diff
@@ -12,10 +12,10 @@ def encode(self, string: str) -> list[int]:
         ...
 
     @abstractmethod
-    def decode(self, encoded: list[int]):
+    def decode(self, encoded: list[int]) -> str:
         ...
 
-    def num_tokens(self, string: str, return_encoded: bool = False):
+    def num_tokens(self, string: str, return_encoded: bool = False) -> int:
```
Comment (hbertrand) [inline buster/tokenizers/base.py:18]: should be `Union[int, tuple[int, str]]` no?

File: buster/busterbot.py (hunk 4 above)
Comment (jerpint) [reply buster/busterbot.py:109]: you are right, i added it

File: buster/tokenizers/base.py (hunk 9 above)
Comment (jerpint) [reply buster/tokenizers/base.py:18]: ah youre right

**Pull request merged**